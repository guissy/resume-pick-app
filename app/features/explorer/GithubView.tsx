import React from 'react';
import { remote } from 'electron';
import path from 'path';
import { GhOpt, HTMLWebview } from '../type';
import styles from './ScoreList.css';
import parseRepo from '../parser/parseRepo';

type Props = {
  url: string;
};
const GithubView: React.FC<Props> = ({ url }) => {
  const webviewRef = React.useRef<
    (HTMLWebview & { loadURL: (url: string) => void }) | null
  >(null);
  const [repos, setRepos] = React.useState(NaN);
  const [contrib, setContrib] = React.useState(NaN);
  const [followers, setFollowers] = React.useState(NaN);
  const tail = url.split('://').pop();
  const [host, pathname] = tail?.split('/') || [];
  const urlOk = `https://${host}/${pathname.split('?').shift()}`;
  const [status, setStatus] = React.useState('prepare...');
  const [repoInfo, setRepoInfo] = React.useState('');
  React.useEffect(() => {
    webviewRef.current?.addEventListener('dom-ready', async () => {
      if (status === 'prepare...') {
        setStatus('loading...');
        const data = await webviewRef.current?.executeJavaScript(
          `JSON.stringify({
            repos: parseInt(document.querySelector('.Counter')?.textContent),
            followers: parseInt(document.querySelector(".js-profile-editable-area a")?.textContent?.trim()),
            contrib: parseInt(document.querySelector('.js-yearly-contributions h2')?.textContent?.trim()),
            calendar: "document.querySelector('.js-calendar-graph-svg').outerHTML",
          })`,
          false
        );
        const result = JSON.parse(data || '{}') as GhOpt;
        if (typeof result.repos === 'number') {
          setRepos(result.repos);
        }
        if (typeof result.followers === 'number') {
          setFollowers(result.followers);
        }
        if (typeof result.contrib === 'number') {
          setContrib(result.contrib);
          if (result.contrib > 5) {
            await webviewRef.current?.executeJavaScript(
              `document.querySelector('a[href$="?tab=repositories"]')?.click()`,
              false
            );
            await new Promise((resolve) => setTimeout(resolve, 3000));
            const json = await webviewRef.current?.executeJavaScript(
              `JSON.stringify(Array.from(document
              .querySelectorAll("#user-repositories-list > ul > li"))
              .map(li => {
                const href = li.querySelector('[itemprop*="codeRepository"]')?.href;
                const commits = Array.from(li.querySelector('svg polyline')?.points || [])
                .map(p => p.y)
                .reduce((s, v) => s + v - 1, 0);
                return [href, commits];
              })
              .sort((a, b)=> a[1] - b[1] > 0 ? -1 : 1))`,
              false
            );
            const top30 = JSON.parse(json || '[]') as [string, number][];
            const first = top30[0]?.[0];
            if (first) {
              const repo = `${first}.git`.replace(
                'github.com',
                'github.com.cnpmjs.org'
              );
              const dir = remote.app.getPath('temp');
              const name = `repo_${Date.now()}`;
              const {
                commits,
                nCode,
                commentRate,
                linesInFile,
                linesInCommit,
              } = await parseRepo(repo, path.join(dir, name), setStatus);
              setRepoInfo(`file ${linesInFile}
commits ${commits.length}
lines ${nCode}
comments ${(commentRate * 100).toFixed(1)}%
commit/line ${linesInCommit}
`);
              setStatus('');
            }
          } else {
            setStatus('');
          }
        }
      }
    });
  }, [webviewRef, status]);

  return ['github.com', 'gitee.com'].includes(host) ? (
    <div>
      <webview
        ref={webviewRef}
        src={urlOk}
        style={{
          width: '100vw',
          opacity: 0.5,
          position: 'absolute',
          zIndex: -1,
          display: 'none',
        }}
      />
      {repos >= 0 && (
        <>
          <span className={styles.subName}>repo</span>
          <span className={styles.gained}>{repos}</span>
          <br />
        </>
      )}
      {followers >= 0 && (
        <>
          <span className={styles.subName}>followers</span>
          <span className={styles.gained}>{followers}</span>
          <br />
        </>
      )}
      {contrib >= 0 && (
        <>
          <span className={styles.subName}>contrib</span>
          <span className={styles.gained}>{contrib}</span>
        </>
      )}
      <br />
      <span style={{ color: '#444444', display: 'block', fontSize: '0.8rem' }}>
        {status}
      </span>
      {repoInfo?.split('\n').map((s) => (
        <span key={s} style={{ color: '#eeeeee', display: 'block' }}>
          {s}
        </span>
      ))}
      {/* <span dangerouslySetInnerHTML={{ __html: calendar }} /> */}
    </div>
  ) : (
    <></>
  );
};

export default GithubView;
