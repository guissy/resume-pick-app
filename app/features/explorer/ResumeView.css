.wrap {
    color: #eeeeee;
    background: #333333;
}

.article {
    height: calc(100vh - 120px);
    overflow: auto;
    padding: 20px 20px 100px 20px;
}
mark {
    border-radius: 4px;
    padding: 0 3px;
    display: inline-block;
    color: #ffffff;
    background-color: #999999;
}

:global(.kw0) {
    background: #e10000;
}
:global(.kw1) {
    background: #e1e100;
}
:global(.kw2) {
    background: #00e100;
}
:global(.kws) {
    background: #ff00ff;
}
:global(.kw0.pos_2_0),
:global(.kw0.pos_1_9),
:global(.kw0.pos_1_8),
:global(.kw0.pos_1_7),
:global(.kw0.pos_1_6) {
    background: rgba(225, 0, 0, 0.8);
}
:global(.kw0.pos_1_5),
:global(.kw0.pos_1_4),
:global(.kw0.pos_1_3),
:global(.kw0.pos_1_2),
:global(.kw0.pos_1_1) {
    background: rgba(195, 0, 0, 0.6);
}
:global(.kw0.pos_1_0),
:global(.kw0.pos_0_9),
:global(.kw0.pos_0_8),
:global(.kw0.pos_0_7),
:global(.kw0.pos_0_6) {
    background: rgba(165, 0, 0, 0.4);
}
:global(.kw0.pos_0_5),
:global(.kw0.pos_0_4),
:global(.kw0.pos_0_3),
:global(.kw0.pos_0_2),
:global(.kw0.pos_0_1) {
    background: rgba(135, 0, 0, 0.2);
}
:global(.kw1.pos_2_0),
:global(.kw1.pos_1_9),
:global(.kw1.pos_1_8),
:global(.kw1.pos_1_7),
:global(.kw1.pos_1_6) {
    background: rgba(225, 225, 0, 0.8);
}
:global(.kw1.pos_1_5),
:global(.kw1.pos_1_4),
:global(.kw1.pos_1_3),
:global(.kw1.pos_1_2),
:global(.kw1.pos_1_1) {
    background: rgba(195, 195, 0, 0.6);
}
:global(.kw1.pos_1_0),
:global(.kw1.pos_0_9),
:global(.kw1.pos_0_8),
:global(.kw1.pos_0_7),
:global(.kw1.pos_0_6) {
    background: rgba(165, 165, 0, 0.4);
}
:global(.kw1.pos_0_5),
:global(.kw1.pos_0_4),
:global(.kw1.pos_0_3),
:global(.kw1.pos_0_2),
:global(.kw1.pos_0_1) {
    background: rgba(135, 135, 0, 0.2);
}
:global(.kw0.pos_2_0),
:global(.kw0.pos_1_9),
:global(.kw0.pos_1_8),
:global(.kw0.pos_1_7),
:global(.kw0.pos_1_6) {
    background: rgba(0, 225, 0, 0.8);
}
:global(.kw2.pos_1_5),
:global(.kw2.pos_1_4),
:global(.kw2.pos_1_3),
:global(.kw2.pos_1_2),
:global(.kw2.pos_1_1) {
    background: rgba(0, 195, 0, 0.6);
}
:global(.kw2.pos_1_0),
:global(.kw2.pos_0_9),
:global(.kw2.pos_0_8),
:global(.kw2.pos_0_7),
:global(.kw2.pos_0_6) {
    background: rgba(0, 165, 0, 0.4);
}
:global(.kw2.pos_0_5),
:global(.kw2.pos_0_4),
:global(.kw2.pos_0_3),
:global(.kw2.pos_0_2),
:global(.kw2.pos_0_1) {
    background: rgba(0, 135, 0, 0.2);
}
:global(.kw2.neg_1_5),
:global(.kw2.neg_1_4),
:global(.kw2.neg_1_3),
:global(.kw2.neg_1_2),
:global(.kw2.neg_1_1) {
    background: rgba(160, 160, 160, 0.8);
}
:global(.kw2.neg_1_0),
:global(.kw2.neg_0_9),
:global(.kw2.neg_0_8),
:global(.kw2.neg_0_7),
:global(.kw2.neg_0_6) {
    background: rgba(160, 160, 160, 0.4);
}
:global(.kw2.neg_0_5),
:global(.kw2.neg_0_4),
:global(.kw2.neg_0_3),
:global(.kw2.neg_0_2),
:global(.kw2.neg_0_1) {
    background: rgba(160, 160, 160, 0.2);
}

.searchWrap {
    display: inline-flex;
    align-items: center;
    margin-left: 20px;
}
.search {
    text-indent: 20px;
    color: #fff;
    width: 200px;
    font-size: 16px;
}
.searchWrap > i {
    transform: translateX(200%) scale(0.8);
}
.count {
    display: inline-block;
    font-size: 8px;
    color: #999999;
    border: 1px solid #545454;
    border-radius: 50%;
    width: 14px;
    height: 14px;
    line-height: 14px;
    text-align: center;
    transform: translateX(-50%);
}
