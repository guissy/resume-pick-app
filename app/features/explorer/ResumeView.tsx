import React, { useEffect } from 'react';
import { shell } from 'electron';
import { useKey } from 'react-use';
// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore
import Highlighter from 'react-highlight-words';
import { ScoreFile } from '../type';
import styles from './ResumeView.css';

type Props = {
  resume: ScoreFile | undefined;
  onClose: () => void;
};
const ResumeView: React.FC<Props> = ({ resume, onClose }) => {
  const topElmRef = React.useRef<HTMLParagraphElement>(null);
  useEffect(() => {
    if (resume && topElmRef.current) {
      topElmRef.current.scrollIntoView({ behavior: 'smooth' });
    }
  }, [resume]);
  const clickOpenNative = React.useCallback(() => {
    if (resume?.path) {
      if (resume?.path.startsWith('http')) {
        shell.openExternal(resume?.path);
      } else {
        shell.openItem(resume.path);
      }
    }
  }, [resume]);
  const [searchTxt, setSearchTxt] = React.useState('');
  const keywords = React.useMemo(
    () =>
      (resume?.keywords || [])
        .map((k) => k.children.map((w) => [w.name, ...(w.alias || [])]))
        .flat(2)
        .concat(searchTxt.split(' ')),
    [resume, searchTxt]
  );
  const count = React.useMemo(() => {
    const searchs = searchTxt?.split(' ').filter(Boolean);
    return searchTxt.trim()
      ? searchs?.reduce((s, v) => {
          const matches = resume?.text.match(new RegExp(v, 'gi')) || [];
          return s + matches.length;
        }, 0)
      : 0;
  }, [searchTxt, resume]);
  const keywordClassName = React.useMemo(
    () =>
      (resume?.keywords || [])
        .map((k, i) =>
          k.children.map((w) =>
            [w.name, ...(w.alias || [])].map((n) => ({
              [n]: `kw${i} ${w.score > 0 ? 'pos' : 'neg'}_${(w.gained / w.score)
                .toFixed(1)
                .replace('.', '_')}`,
            }))
          )
        )
        .flat(2)
        .concat(searchTxt.split(' ').map((s) => ({ [s]: 'kws' })))
        .reduce((s, v) => ({ ...s, ...v }), {}),
    [resume, searchTxt]
  );
  useKey('Escape', onClose);

  return (
    <div className={styles.wrap}>
      <header style={{ textAlign: 'right' }}>
        <label htmlFor="search" className={styles.searchWrap}>
          <i className="fa fa-search" />
          <span
            className={styles.count}
            style={{ visibility: count > 0 ? 'visible' : 'hidden' }}
          >
            {count}
          </span>
          <input
            className={styles.search}
            id="search"
            type="search"
            value={searchTxt}
            placeholder="输入关键字搜索简历"
            onChange={(e) => {
              setSearchTxt(e.target.value);
            }}
          />
        </label>
        <button
          type="button"
          onClick={clickOpenNative}
          style={{
            background: 'none',
            border: '1px solid #fff',
            color: '#fff',
            fontSize: 14,
            cursor: 'pointer',
          }}
        >
          使用本地软件打开此文件
        </button>
        <button type="button" onClick={onClose}>
          <i className="fa fa-times fa-2x" />
        </button>
      </header>
      <article className={styles.article} key={searchTxt}>
        {(resume?.text || '').split('\n').map((content, i) => (
          <p
            key={String(i)}
            style={{ marginBlockStart: 5, marginBlockEnd: 5 }}
            ref={i === 0 ? topElmRef : undefined}
          >
            <Highlighter
              highlightClassName={keywordClassName}
              searchWords={keywords}
              textToHighlight={content}
            />
          </p>
        ))}
      </article>
    </div>
  );
};

export default ResumeView;
