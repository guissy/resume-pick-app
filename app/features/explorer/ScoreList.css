.tableWrap {
  padding: 0 20px 30px 20px;
  color: #aaa;
}

.header {
  padding-top: 10px;
  padding-bottom: 10px;
}

.table {
  width: 100%;
  text-align: center;
  border-spacing: 0;
  border-top: 1px solid #666;
  border-left: 1px solid #666;
  box-sizing: border-box;
}

.table > thead {
  color: #aaa;
}

.td {
  padding: 5px;
  border-right: 1px solid #666;
  border-bottom: 1px solid #666;
}

.score {
  padding: 5px 10px 5px 5px;
  border-right: 1px solid #666;
  border-bottom: 1px solid #666;
  text-align: right;
  width: 50px;
}

.scoreItem {
  display: flex;
  justify-content: space-around;
  column-gap: 20px;
  gap: 20px;
  line-height: 1.4rem;
}

.subName {
  display: inline-block;
  padding-right: 6px;
  text-align: right;
  flex: 3;
}

.gained {
  display: inline-block;
  padding-left: 6px;
  text-align: left;
  flex: 2;
}

.level {
  position: relative;
  display: inline-block;
  margin: 0 auto 10px;
  height: 38px;
  line-height: 38px;
  min-width: 30px;
  padding: 0 4px;
  top: 5px;
  border-radius: 21px;
  text-align: center;
  font-size: 15px;
  color: #fffab8;
  text-shadow: 0 0 3px rgba(255, 215, 0, 1);
  background-color: rgba(33, 33, 33, 0.6);
  border: 1px solid #8e8410;
  text-transform: capitalize;
}
.level > i {
  position: absolute;
  top: 15px;
  width: 10px;
  font-size: 8px;
}
.level > i:first-child {
  right: -1px;
}
.level > i:nth-child(2) {
  right: -8px;
}

.level9,
.level8,
.level7 {
  color: #fffab8;
  font-size: 19px;
  font-weight: bolder;
  border: 1px solid #ccbe17;
}
.level6 {
  color: #ded99f;
  text-shadow: 0 0 2px rgba(255, 215, 0, 1);
  font-size: 17px;
}
.level5 {
  color: #bab789;
  text-shadow: unset;
  border: 1px solid #5c560b;
}
.level4 {
  color: #9f9c83;
  text-shadow: unset;
  border: 1px solid #302d06;
}

.level3 {
  color: #7b7b7a;
  text-shadow: unset;
  border: 1px solid #272705;
}
.level2 {
  color: #636363;
  text-shadow: unset;
  border: 1px solid #272705;
}
.level1 {
  color: #545454;
  text-shadow: unset;
  border: 1px solid #272705;
}

.link {
  border: 0;
  padding: 0;
  background: none;
  font-size: 14px;
  color: #aaa;
  cursor: pointer;
}

.link:hover {
  color: #8989ff;
}

.namePhone {
  display: block;
}

.alink {
  display: flex;
  align-items: center;
  font-size: 14px;
  cursor: pointer;
  overflow: hidden;
  width: 150px;
  white-space: nowrap;
  text-overflow: ellipsis;
  margin: 0 auto;
}
.blogIcon {
  display: inline-block;
  height: 12px;
  margin-right: 5px;
}

.footer {
  padding: 10px 0;
  text-align: right;
}

.exportExcel {
  background: none;
  font-size: 14px;
  color: #666;
  border-color: #666;
  cursor: pointer;
}
.exportExcel:hover {
  border-color: #ddd;
  color: #ddd;
}

.sort {
  position: relative;
  cursor: pointer;
}

.sortBtn {
  display: block;
  background: none;
  color: #aaa;
  border: 0;
  font-size: unset;
  cursor: pointer;
}

.sortBtn:not(:focus-visible) {
  outline: none;
}

.caretInit {
  top: 50%;
  position: absolute;
  display: inline-flex;
  flex-direction: column;
  transform: translate(2px, -50%);
  color: #666;
}

.active {
  color: #fff;
}

.caretInit > i {
  line-height: 8px;
}
ul.icons {
  padding-left: 0;
  margin-top: 10px;
  margin-bottom: 0;
}
.icons > li {
  display: inline-flex;
  justify-content: center;
  align-items: center;
  width: 32px;
  height: 32px;
  border-radius: 50%;
  background: #eee;
  margin-right: 5px;
  margin-bottom: 5px;
}
.icons > li:nth-child(6n) {
  margin-right: 0;
}
.icons img {
  height: 80%;
}
.searchWrap {
  display: inline-flex;
  align-items: center;

  /* margin-left: 20px; */
}
.search {
  text-indent: 20px;
  color: #fff;
  width: 200px;
  font-size: 16px;
}
.searchWrap i {
  transform: translateX(140%) scale(0.8);
}

.opts {
  display: flex;
  align-items: center;
}
header > label {
  cursor: pointer;
  margin-right: 10px;
  color: #999;
}
.opts > label {
  color: #999;
  margin-right: 5px;
}
header > label > input,
.opts > label > input {
  margin-right: 5px;
}

.opt {
  display: inline-block;
  width: 100%;
  height: 100%;
  background: none;
  border: 0;
  padding: 0;
  margin: 0;
  text-align: center;
}

.trash {
  background: none;
  border: 0;
  padding: 0;
  color: #666;
  cursor: pointer;
  font-size: 12px;
}
.trash:hover {
  color: #f00;
}
.trash > span {
  padding-right: 5px;
}
