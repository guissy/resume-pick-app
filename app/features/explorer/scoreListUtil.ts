import styles from './ScoreList.css';

export function toggleSort([up, down]: [string, string]) {
  return (sorting: string) => {
    if (sorting !== up && sorting !== down) {
      return down;
    }
    if (sorting === up) {
      return '';
    }
    if (sorting === down) {
      return up;
    }
    return '';
  };
}

export function getLevelStyle(level: string) {
  const classNames = [
    styles.level1,
    styles.level2,
    styles.level3,
    styles.level4,
    styles.level5,
    styles.level6,
    styles.level7,
    styles.level8,
    styles.level9,
  ];
  const [n] = level?.match(/\d/g) || ['1'];
  const index = parseInt(n, 10);
  if (index > 0 && index <= classNames.length) {
    return `${styles.level} ${classNames[index - 1]}`;
  }
  return '';
}

export default {
  getLevelStyle,
};
