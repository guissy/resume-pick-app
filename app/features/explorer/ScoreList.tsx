import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { shell } from 'electron';
import { useDebounce } from 'react-use';
import uniqBy from 'lodash/uniqBy';
import styles from './ScoreList.css';
import {
  selectNameScore,
  removeNameScore,
  clearNameScore,
} from '../scoreSlice';
import { getBlogByLink, getGithubByLink } from '../parser/tractWorkAge';
import { selectConfig, updateConfig } from '../configSlice';
import Image, { imgKey } from '../icon/image';
import GithubView from './GithubView';
import { ScoreFile, SortKey } from '../type';
import TreeMap from './TreeMap';
import ResumeView from './ResumeView';
import exportExcel from '../parser/exportExcel';

type File = ScoreFile | undefined;
type Props = {
  search: string;
  setSearch: (kw: string) => void;
};

function getLevelStyle(level: string) {
  const classNames = [
    styles.level1,
    styles.level2,
    styles.level3,
    styles.level4,
    styles.level5,
    styles.level6,
    styles.level7,
  ];
  const [n] = level?.match(/\d/g) || ['1'];
  const index = parseInt(n, 10);
  if (index > 0 && index <= classNames.length) {
    return `${styles.level} ${classNames[index - 1]}`;
  }
  return '';
}

const ScoreList: React.FC<Props> = ({ search, setSearch }) => {
  const scores = useSelector(selectNameScore);
  const config = useSelector(selectConfig);
  const dispatch = useDispatch();
  const onClickExport = React.useCallback(() => {
    exportExcel(config, scores);
  }, [config, scores]);
  const [sort, setSort] = React.useState<SortKey>('');
  const onClickWorkAge = React.useCallback(() => {
    setSort((wa) => {
      if (wa !== 'workAgeUp' && wa !== 'workAgeDown') {
        return 'workAgeDown';
      }
      if (wa === 'workAgeUp') {
        return '';
      }
      if (wa === 'workAgeDown') {
        return 'workAgeUp';
      }
      return '';
    });
  }, []);
  const onClickScore = React.useCallback(() => {
    setSort((wa) => {
      if (wa !== 'scoreUp' && wa !== 'scoreDown') {
        return 'scoreDown';
      }
      if (wa === 'scoreUp') {
        return '';
      }
      if (wa === 'scoreDown') {
        return 'scoreUp';
      }
      return '';
    });
  }, []);
  const optionDefault = React.useMemo(() => {
    return imgKey.filter((img) =>
      (config || [])
        .map((c) => c.children.map((w) => w.name))
        .flat()
        .includes(img)
    );
  }, [config]);
  const [cacheScore] = React.useState(new Map<string, number>()); // 缓存大于0
  const [showFull, setShowFull] = React.useState(true);
  const [gitRepo, setGitRepo] = React.useState(false);
  const [option, setOption] = React.useState(
    new Map(optionDefault.map((opt) => [opt, true]))
  );
  React.useEffect(() => {
    if (Array.isArray(config)) {
      setOption((opt) => {
        config.forEach((k) => {
          k.children.forEach((w) => {
            if (optionDefault.includes(w.name)) {
              opt.set(w.name, w.score !== 0);
              if (w.score !== 0) {
                cacheScore.set(w.name, w.score);
              }
            }
          });
        });
        return new Map(opt);
      });
    }
  }, [optionDefault, cacheScore, config]);
  const [checkedAll, setCheckedAll] = React.useState(true);
  const onClickOpt = React.useCallback(
    (opt: string) => {
      option.set(opt, !option.get(opt));
      setOption(new Map(option));
      const configOk = config.map((k) => ({
        ...k,
        children: k.children.map((w) => {
          let score = w.score || 0;
          if (option.has(w.name)) {
            score = option.get(w.name) ? cacheScore.get(w.name) || score : 0;
          }
          return {
            ...w,
            score,
          };
        }),
      }));
      setCheckedAll(false);
      dispatch(updateConfig({ default: configOk }));
    },
    [option, config, cacheScore, dispatch]
  );
  const onClickCheckAll = React.useCallback(() => {
    setCheckedAll(!checkedAll);
    if (config) {
      setOption(new Map(optionDefault.map((opt) => [opt, true])));
      const configOk = config.map((k) => ({
        ...k,
        children: k.children.map((w) => {
          let score = w.score || 0;
          if (option.has(w.name)) {
            score = !checkedAll ? cacheScore.get(w.name) || w.score : 0;
          }
          return {
            ...w,
            score,
          };
        }),
      }));
      dispatch(updateConfig({ default: configOk }));
    }
  }, [checkedAll, config, cacheScore, option, optionDefault, dispatch]);
  const onClickLink = React.useCallback((e, url) => {
    e.preventDefault();
    const urlOk = url.startsWith('http') ? url : `http://${url}`;
    shell.openExternal(urlOk);
  }, []);

  const [showDialog, setShowDialog] = React.useState(false);
  const [resumeActive, setResumeActive] = React.useState<File>();
  const onOpenResume = React.useCallback((scoreFile: ScoreFile) => {
    setShowDialog(true);
    setResumeActive(scoreFile);
  }, []);
  const onCloseResume = React.useCallback(() => {
    setShowDialog(false);
    setResumeActive(undefined);
  }, []);
  const [searchTxt, setSearchTxt] = React.useState(search);
  useDebounce(() => setSearch(searchTxt), 2000, [searchTxt]);
  const onClear = React.useCallback(() => {
    dispatch(clearNameScore());
  }, [dispatch]);
  const onRemove = React.useCallback(
    (path) => {
      dispatch(removeNameScore(path));
    },
    [dispatch]
  );
  return scores.length > 0 ? (
    <div role="presentation" className={styles.tableWrap}>
      <dialog
        open={showDialog}
        style={{
          position: 'fixed',
          padding: 0,
          zIndex: 100,
          width: '80vw',
          top: 10,
        }}
      >
        <ResumeView resume={resumeActive} onClose={onCloseResume} />
      </dialog>
      <header className={styles.header}>
        <label htmlFor="showFull">
          <input
            id="showFull"
            type="checkbox"
            onChange={() => setShowFull(!showFull)}
            checked={showFull}
          />
          Show Full
        </label>
        <label htmlFor="gitRepo">
          <input
            id="gitRepo"
            type="checkbox"
            onChange={() => setGitRepo(!gitRepo)}
            checked={gitRepo}
          />
          Git Repo
        </label>
        <button className={styles.trash} onClick={onClear} type="button">
          <span>清空</span>
          <i className="fa fa-trash" />
        </button>
        <label htmlFor="search" className={styles.searchWrap}>
          <i className="fa fa-search" />
          <input
            className={styles.search}
            id="search"
            type="search"
            value={searchTxt}
            placeholder="输入关键字搜索简历"
            onChange={(e) => {
              setSearchTxt(e.target.value);
            }}
          />
        </label>
        <div className={styles.opts}>
          <label htmlFor="optCheckbox">
            <input
              id="optCheckbox"
              type="checkbox"
              onChange={onClickCheckAll}
              checked={checkedAll}
            />
            全选
          </label>
          <ul className={styles.icons}>
            {config
              .map((k) => k.children.map((w) => w.name))
              .flat()
              .filter((w) => option.has(w))
              .map((w) => (
                <li
                  key={w}
                  style={{
                    backgroundColor: option.get(w) ? '#333' : '#aaa',
                  }}
                >
                  <button
                    type="button"
                    onClick={() => onClickOpt(w)}
                    className={styles.opt}
                    style={{
                      filter: option.get(w) ? 'unset' : 'grayscale(0.9)',
                    }}
                  >
                    <img
                      src={Image[`${w}_png` as keyof typeof Image]}
                      alt={w}
                    />
                  </button>
                </li>
              ))}
          </ul>
        </div>
      </header>
      <table className={styles.table}>
        <thead>
          <tr>
            <td className={styles.td} style={{ width: 32 }}>
              序号
            </td>
            <td className={styles.td} style={{ width: 260 }}>
              文件
            </td>
            <td className={`${styles.td} ${styles.sort}`} style={{ width: 50 }}>
              <button
                className={styles.sortBtn}
                type="button"
                onClick={onClickWorkAge}
              >
                <span>经验</span>
                <div className={styles.caretInit}>
                  <i
                    className={`fa fa-caret-up ${
                      sort === 'workAgeUp' ? styles.active : ''
                    }`}
                  />
                  <i
                    className={`fa fa-caret-down ${
                      sort === 'workAgeDown' ? styles.active : ''
                    }`}
                  />
                </div>
              </button>
            </td>
            <td className={`${styles.td} ${styles.sort}`}>
              <button
                className={styles.sortBtn}
                type="button"
                onClick={onClickScore}
              >
                <span>分数</span>
                <div className={styles.caretInit}>
                  <i
                    className={`fa fa-caret-up ${
                      sort === 'scoreUp' ? styles.active : ''
                    }`}
                  />
                  <i
                    className={`fa fa-caret-down ${
                      sort === 'scoreDown' ? styles.active : ''
                    }`}
                  />
                </div>
              </button>
            </td>
            <td colSpan={1} className={styles.td} style={{ width: 150 }}>
              github
            </td>
            <td colSpan={1} className={styles.td} style={{ width: 'auto' }}>
              关键字
            </td>
          </tr>
        </thead>
        <tbody>
          {scores
            .map((s) => ({ ...s }))
            .sort((a, b) => {
              if (sort === 'workAgeDown') {
                return a.workAge > b.workAge ? -1 : 1;
              }
              if (sort === 'workAgeUp') {
                return a.workAge < b.workAge ? -1 : 1;
              }
              if (sort === 'scoreDown') {
                return a.score > b.score ? -1 : 1;
              }
              if (sort === 'scoreUp') {
                return a.score < b.score ? -1 : 1;
              }
              if (sort === 'sentimentDown') {
                return a.sentiment > b.sentiment ? -1 : 1;
              }
              if (sort === 'sentimentUp') {
                return a.sentiment < b.sentiment ? -1 : 1;
              }
              return a.levelValue > b.levelValue ? -1 : 1;
            })
            .map((v, i) => (
              <tr key={v.path}>
                <td className={styles.td}>
                  {i + 1}
                  <br />
                  <button
                    className={styles.trash}
                    onClick={() => onRemove(v.path)}
                    type="button"
                  >
                    <i className="fa fa-trash" />
                  </button>
                </td>
                <td className={styles.td}>
                  <button
                    type="button"
                    className={styles.link}
                    onClick={(e) => {
                      e.stopPropagation();
                      onOpenResume(v);
                    }}
                  >
                    {showFull && (
                      <>
                        <span className={styles.namePhone}>{v.name}</span>
                        <span className={styles.namePhone}>{v.phone}</span>
                        <span className={styles.namePhone}>{v.school}</span>
                      </>
                    )}
                    <ul className={styles.icons}>
                      {uniqBy(
                        v.keywords
                          .map((k) => k.children.map((w) => w))
                          .flat()
                          .filter((w) =>
                            Array.from(option.entries())
                              .filter(([_k, b]) => _k && b)
                              .map(([k]) => k)
                              .includes(w.name)
                          ),
                        'name'
                      ).map((w) => (
                        <li
                          key={w.name}
                          style={{
                            backgroundColor: `rgba(235, 235, 235, ${
                              w.gained / w.score + 0.2
                            })`,
                          }}
                        >
                          <img
                            src={Image[`${w.name}_png` as keyof typeof Image]}
                            style={{ opacity: w.gained / w.score }}
                            alt={w.name}
                          />
                        </li>
                      ))}
                    </ul>
                  </button>
                </td>
                <td className={styles.td}>
                  {v.workAge > 0 ? `${Math.round(v.workAge * 2) / 2}年` : '-'}
                </td>
                <td className={styles.score}>{v.score?.toFixed(2)}</td>
                <td className={styles.td}>
                  <span
                    className={getLevelStyle(v.level)}
                    title={String(v.levelValue)}
                  >
                    {v.level}
                  </span>
                  {gitRepo &&
                    getGithubByLink(v.links).map((link) => (
                      <GithubView key={link} url={link} />
                    ))}
                  {v.links.map((link) => (
                    <a
                      key={link}
                      href={link}
                      className={styles.alink}
                      style={{
                        color: getBlogByLink(link) ? '#f1f1f1' : '#999999',
                      }}
                      onClick={(e) => onClickLink(e, link)}
                    >
                      <img
                        className={styles.blogIcon}
                        src={
                          Image[
                            `${getBlogByLink(link)}_png` as keyof typeof Image
                          ]
                        }
                        alt={getBlogByLink(link)}
                      />
                      {link.replace(/(https?:\/\/|www\.)/, '')}
                    </a>
                  ))}
                </td>
                <td className={styles.td}>
                  <TreeMap scoreFile={v} search={search} />
                </td>
              </tr>
            ))}
        </tbody>
      </table>
      <footer className={styles.footer}>
        <button
          className={styles.exportExcel}
          type="button"
          onClick={onClickExport}
        >
          导出 Excel
        </button>
      </footer>
    </div>
  ) : (
    <></>
  );
};

export default ScoreList;
