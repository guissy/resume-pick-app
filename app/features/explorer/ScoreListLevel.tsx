import React from 'react';
import { getLevelStyle } from './scoreListUtil';

type Props = {
  level: string;
  levelValue: number;
};

const ScoreListLevel: React.FC<Props> = ({ level, levelValue }) => {
  return (
    <span className={getLevelStyle(level)} title={String(levelValue)}>
      {level?.slice(0, 2)}
      {level
        ?.split('')
        .filter((c) => c === '+')
        .map((c, n) => (
          <i key={c + String(n)} className="fa fa-plus fa-1x" title={c} />
        ))}
    </span>
  );
};

export default ScoreListLevel;
