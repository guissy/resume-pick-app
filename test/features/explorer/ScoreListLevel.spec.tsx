import React from 'react';
import Enzyme, { mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import renderer from 'react-test-renderer';
import ScoreListLevel from '../../../app/features/explorer/ScoreListLevel';
import * as scoreListUtil from '../../../app/features/explorer/scoreListUtil';

Enzyme.configure({ adapter: new Adapter() });

const defaultTestProps = { level: 'p6++', levelValue: 7 };

function setup(props = defaultTestProps) {
  // eslint-disable-next-line react/jsx-props-no-spreading
  const getWrapper = () => mount(<ScoreListLevel {...props} />);
  const component = getWrapper();
  return {
    component,
  };
}
describe('ScoreListLevel component', () => {
  it('should match exact snapshot', () => {
    const tree = renderer
      // eslint-disable-next-line react/jsx-props-no-spreading
      .create(<ScoreListLevel {...defaultTestProps} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should show level', () => {
    const getLevelStyleSpy = jest.spyOn(scoreListUtil, 'getLevelStyle');
    const { component } = setup();
    expect(component.text()).toContain(defaultTestProps.level.slice(0, 2));
    expect(getLevelStyleSpy).toBeCalledWith(defaultTestProps.level);
    expect(getLevelStyleSpy).toBeCalledTimes(1);
  });
});
