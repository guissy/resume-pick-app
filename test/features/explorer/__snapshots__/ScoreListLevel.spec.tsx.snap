// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScoreListLevel component should match exact snapshot 1`] = `
<span
  className="level level6"
  title="7"
>
  p6
  <i
    className="fa fa-plus fa-1x"
    title="+"
  />
  <i
    className="fa fa-plus fa-1x"
    title="+"
  />
</span>
`;
